File: sdks/typescript/core/ai/music-input.ts
import { createNimiError } from '../../types';

export type NimiMusicGenerationInputProfile = {
  readonly lyricsMode: 'unsupported' | 'optional' | 'required';
  readonly scoreMode: 'unsupported' | 'required';
  readonly scoreFormats: readonly ('abc' | 'midi')[];
  readonly scoreConditioning: readonly ('melody-only' | 'melody-and-harmony')[];
  readonly supportsInstrumental: boolean;
  readonly supportsSeed: boolean;
  readonly supportsGeneratedScore: boolean;
  readonly supportsAudioReference: boolean;
  readonly maxDurationSeconds: number;
  readonly defaultDurationSeconds: number;
  readonly maxPromptBytes: number;
  readonly maxLyricsBytes: number;
  readonly maxScoreBytes: number;
  readonly maxAudioReferenceBytes: number;
};
export type NimiMusicTranscriptionInputProfile = {
  readonly formats: readonly ('abc' | 'midi' | 'timeline')[];
  readonly parts: readonly ('vocal-melody' | 'lead-sheet' | 'full-arrangement')[];
  readonly maxDurationSeconds: number;
  readonly maxSourceBytes: number;
  readonly supportsRange: boolean;
};
export type NimiMusicInputCapabilities = {
  readonly generation: readonly NimiMusicGenerationInputProfile[];
  readonly transcription?: readonly NimiMusicTranscriptionInputProfile[];
};

const fail = (): never => { throw createNimiError({ reasonCode: 'SDK_LOCAL_APP_PROJECTION_INVALID', message: 'Music input capabilities are invalid.', actionHint: 'update_matching_runtime_sdk_kit', source: 'sdk' }); };
const record = (value: unknown): Record<string, unknown> => {
  if (!value || typeof value !== 'object' || Array.isArray(value)) fail();
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail();
  return value as Record<string, unknown>;
};
const integer = (value: unknown, min: number, max: number): value is number => typeof value === 'number' && Number.isSafeInteger(value) && value >= min && value <= max;
const tokens = (value: unknown, allowed: readonly string[]): value is string[] => Array.isArray(value) && value.length <= allowed.length && new Set(value).size === value.length && value.every((item) => typeof item === 'string' && allowed.includes(item));

// @nimi-authority: rule.nimi.runtime.ai-provider.music-generation
export function projectMusicInputCapabilities(value: unknown): NimiMusicInputCapabilities {
  const source = record(value);
  if (Object.keys(source).some(key => !['generation', 'transcription'].includes(key)) || !Array.isArray(source.generation) || source.generation.length > 16
    || (source.transcription !== undefined && (!Array.isArray(source.transcription) || source.transcription.length > 16))) fail();
  const generation = (source.generation as unknown[]).map((entry) => {
    const row = record(entry);
    const keys = ['lyricsMode', 'scoreMode', 'scoreFormats', 'scoreConditioning', 'supportsInstrumental', 'supportsSeed', 'supportsGeneratedScore', 'supportsAudioReference', 'maxDurationSeconds', 'defaultDurationSeconds', 'maxPromptBytes', 'maxLyricsBytes', 'maxScoreBytes', 'maxAudioReferenceBytes'];
    if (Object.keys(row).length !== keys.length || Object.keys(row).some((key) => !keys.includes(key))
      || !['unsupported', 'optional', 'required'].includes(String(row.lyricsMode)) || !['unsupported', 'required'].includes(String(row.scoreMode))
      || !tokens(row.scoreFormats, ['abc', 'midi']) || !tokens(row.scoreConditioning, ['melody-only', 'melody-and-harmony'])
      || ['supportsInstrumental', 'supportsSeed', 'supportsGeneratedScore', 'supportsAudioReference'].some((key) => typeof row[key] !== 'boolean')
      || !integer(row.maxDurationSeconds, 1, 600) || !integer(row.defaultDurationSeconds, 1, Number(row.maxDurationSeconds))
      || !integer(row.maxPromptBytes, 1, 32768) || !integer(row.maxLyricsBytes, 0, 32768)
      || !integer(row.maxScoreBytes, 0, 1048576) || !integer(row.maxAudioReferenceBytes, 0, 33554432)) fail();
    const profile = row as unknown as NimiMusicGenerationInputProfile;
    if ((profile.scoreMode === 'required') !== (profile.scoreFormats.length > 0 && profile.scoreConditioning.length > 0 && profile.maxScoreBytes > 0)
      || (profile.scoreMode === 'unsupported' && (profile.scoreFormats.length || profile.scoreConditioning.length || profile.maxScoreBytes))
      || profile.supportsAudioReference !== (profile.maxAudioReferenceBytes > 0)
      || (profile.lyricsMode === 'unsupported') !== (profile.maxLyricsBytes === 0)
      || (profile.lyricsMode === 'required' && profile.supportsInstrumental)) fail();
    return Object.freeze({ ...profile, scoreFormats: Object.freeze([...profile.scoreFormats]), scoreConditioning: Object.freeze([...profile.scoreConditioning]) });
  });
  const transcription = ((source.transcription ?? []) as unknown[]).map(entry => {
    const row = record(entry); const keys = ['formats', 'parts', 'maxDurationSeconds', 'maxSourceBytes', 'supportsRange'];
    if (Object.keys(row).length !== keys.length || Object.keys(row).some(key => !keys.includes(key))
      || !tokens(row.formats, ['abc', 'midi', 'timeline']) || !row.formats.length
      || !tokens(row.parts, ['vocal-melody', 'lead-sheet', 'full-arrangement']) || !row.parts.length
      || !integer(row.maxDurationSeconds, 1, 600) || !integer(row.maxSourceBytes, 1, 512 * 1024 * 1024) || typeof row.supportsRange !== 'boolean') fail();
    const profile = row as unknown as NimiMusicTranscriptionInputProfile;
    return Object.freeze({ ...profile, formats: Object.freeze([...profile.formats]), parts: Object.freeze([...profile.parts]) });
  });
  if (!generation.length && !transcription.length) fail();
  return Object.freeze({ generation: Object.freeze(generation), ...(transcription.length ? { transcription: Object.freeze(transcription) } : {}) });
}

File: kit/shell/protected-local/src/windows_local_app/music_input.rs
use serde_json::{json, Value};
use crate::{generated::MusicInputCapabilities, LocalAppOperationError};
use super::untrusted;

// @nimi-authority: rule.nimi.runtime.ai-provider.music-generation
pub(super) fn project(value: MusicInputCapabilities) -> Result<Value, LocalAppOperationError> {
    if value.generation.len() > 16 || value.transcription.len() > 16 || value.voice_convert.len() > 16
        || (value.generation.is_empty() && value.transcription.is_empty() && value.voice_convert.is_empty()) { return Err(untrusted()); }
    let mut profiles = Vec::new();
    for row in value.generation {
        if !matches!(row.lyrics_mode.as_str(), "unsupported" | "optional" | "required")
            || !matches!(row.score_mode.as_str(), "unsupported" | "required")
            || row.max_duration_seconds == 0 || row.max_duration_seconds > 600
            || row.default_duration_seconds == 0 || row.default_duration_seconds > row.max_duration_seconds
            || row.max_prompt_bytes == 0 || row.max_prompt_bytes > 32768 || row.max_lyrics_bytes > 32768
            || row.max_score_bytes > 1048576 || row.max_audio_reference_bytes > 33554432
            || row.supports_audio_reference != (row.max_audio_reference_bytes > 0)
            || (row.lyrics_mode == "unsupported") != (row.max_lyrics_bytes == 0)
            || (row.lyrics_mode == "required" && row.supports_instrumental)
            || !tokens(&row.score_formats, &["abc", "midi"])
            || !tokens(&row.score_conditioning, &["melody-only", "melody-and-harmony"])
            || (row.score_mode == "required") != (!row.score_formats.is_empty() && !row.score_conditioning.is_empty() && row.max_score_bytes > 0)
            || (row.score_mode == "unsupported" && (!row.score_formats.is_empty() || !row.score_conditioning.is_empty() || row.max_score_bytes != 0)) { return Err(untrusted()); }
        profiles.push(json!({"lyricsMode":row.lyrics_mode,"scoreMode":row.score_mode,"scoreFormats":row.score_formats,
            "scoreConditioning":row.score_conditioning,"supportsInstrumental":row.supports_instrumental,"supportsSeed":row.supports_seed,
            "supportsGeneratedScore":row.supports_generated_score,"supportsAudioReference":row.supports_audio_reference,
            "maxDurationSeconds":row.max_duration_seconds,"defaultDurationSeconds":row.default_duration_seconds,
            "maxPromptBytes":row.max_prompt_bytes,"maxLyricsBytes":row.max_lyrics_bytes,"maxScoreBytes":row.max_score_bytes,
            "maxAudioReferenceBytes":row.max_audio_reference_bytes}));
    }
    let mut transcription = Vec::new();
    for row in value.transcription {
        if row.formats.is_empty() || row.parts.is_empty()
            || !tokens(&row.formats, &["abc", "midi", "timeline"])
            || !tokens(&row.parts, &["vocal-melody", "lead-sheet", "full-arrangement"])
            || row.max_duration_seconds == 0 || row.max_duration_seconds > 600
            || row.max_source_bytes == 0 || row.max_source_bytes > 512 * 1024 * 1024 { return Err(untrusted()); }
        transcription.push(json!({"formats":row.formats,"parts":row.parts,"maxDurationSeconds":row.max_duration_seconds,
            "maxSourceBytes":row.max_source_bytes,"supportsRange":row.supports_range}));
    }
    let mut voice_convert = Vec::new();
    for row in value.voice_convert {
        if row.source_kinds.is_empty() || row.target_kinds.is_empty()
            || !tokens(&row.source_kinds, &["singing"])
            || !tokens(&row.target_kinds, &["reference-audio", "preset", "voice-asset"])
            || row.max_source_seconds == 0 || row.max_source_seconds > 600
            || row.max_target_seconds == 0 || row.max_target_seconds > 600
            || row.min_semitone_shift < -12 || row.min_semitone_shift > 12
            || row.max_semitone_shift < -12 || row.max_semitone_shift > 12
            || row.min_semitone_shift > row.max_semitone_shift
            || row.max_source_bytes == 0 || row.max_source_bytes > 512 * 1024 * 1024
            || row.max_target_bytes == 0 || row.max_target_bytes > 512 * 1024 * 1024 { return Err(untrusted()); }
        voice_convert.push(json!({"sourceKinds":row.source_kinds,"targetKinds":row.target_kinds,
            "maxSourceSeconds":row.max_source_seconds,"maxTargetSeconds":row.max_target_seconds,"supportsRange":row.supports_range,
            "supportsSemitoneShift":row.supports_semitone_shift,"minSemitoneShift":row.min_semitone_shift,"maxSemitoneShift":row.max_semitone_shift,
            "maxSourceBytes":row.max_source_bytes,"maxTargetBytes":row.max_target_bytes}));
    }
    let mut result = json!({"generation":profiles});
    if !transcription.is_empty() { result["transcription"] = json!(transcription); }
    if !voice_convert.is_empty() { result["voiceConvert"] = json!(voice_convert); }
    Ok(result)
}

fn tokens(values: &[String], allowed: &[&str]) -> bool {
    values.len() <= allowed.len() && values.iter().enumerate().all(|(index, item)| allowed.contains(&item.as_str()) && !values[..index].contains(item))
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn capability_profiles_keep_only_consistent_input_combinations() {
        let profile = crate::generated::MusicGenerationInputProfile {
            lyrics_mode:"required".into(), score_mode:"unsupported".into(), supports_seed:true,
            supports_generated_score:true, max_duration_seconds:600, default_duration_seconds:20,
            max_prompt_bytes:32768, max_lyrics_bytes:32768, ..Default::default()
        };
        let value = MusicInputCapabilities { generation:vec![profile.clone()], ..Default::default() };
        assert_eq!(project(value.clone()).unwrap()["generation"][0]["maxDurationSeconds"],600);
        let mut contradictory=value.clone();contradictory.generation[0].supports_audio_reference=true;
        assert!(project(contradictory).is_err());
        let mut extra_score=value;extra_score.generation[0].score_formats.push("abc".into());
        assert!(project(extra_score).is_err());
    }
}
